Remove Concat intrinsic in favor of IrStringConcatenation.

sfs opened the pull request:
This PR makes FlattenStringConcatenation apply to nullable strings, which means that all occurrences of String.plus will be represented by IrStringConcatenation. This in turn reduces code duplication in the JVM_IR backend.

sfs commented:
Updated the PR to remove the StringPlus intrinsic, which was also redundant.

sfs commented:
Rebased to current master.

@max-kammerer Is there anything preventing this PR from being merged?

max-kammerer commented:
@sfs Sorry for delay

max-kammerer commented on compiler/testData/codegen/bytecodeText/stringOperations/nullableStringPlus.kt:
```diff
@@ -1,3 +1,6 @@
+// IGNORE_BACKEND: JVM_IR
```
@sfs Please add same bytecode test for IR backend

max-kammerer commented on compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/FlattenStringConcatenationLowering.kt:
```diff
@@ -73,11 +73,11 @@ class FlattenStringConcatenationLowering(val context: CommonBackendContext) : Fi
-                    val dispatchReceiver = expression.dispatchReceiver
-                    dispatchReceiver != null &&
-                            dispatchReceiver.type.isString() &&
+                    val receiver = expression.dispatchReceiver ?: expression.extensionReceiver
```
@sfs Please also check package for irCall, this code would not work correctly for next cases (please also add test for it):
```
operator fun String?.plus(p: String) {
    this
}

fun test(a: String?, b: String): String {
return a + b
}
```

sfs replied:
Thanks, I completely missed this!

udalov commented:
Please ensure there's a test that checks that bytecode is optimized in simple call case such as
```
fun f(s: String?, t: String): String {
    return s.plus(t)
}
```

max-kammerer commented:
@sfs Please rebase on master there could be some affects of #2274

sfs commented:
@max-kammerer @udalov 
Rebased and added more tests.

I changed ExpressionCodegen to generate a call to Intrinsics.stringPlus whenever possibly instead of always using StringBuilder. This way it's compatible with the current tests and produces slightly smaller code.

max-kammerer commented on compiler/testData/codegen/box/strings/simpleStringPlus.kt:
```diff
@@ -0,0 +1,18 @@
+}
+
+fun box(): String {
+    if (f("O", "K") != "OK") return "Fail"
```
Fail 1

sfs replied:
Done

max-kammerer commented on compiler/testData/codegen/box/strings/simpleStringPlus.kt:
```diff
@@ -0,0 +1,18 @@
+
+fun box(): String {
+    if (f("O", "K") != "OK") return "Fail"
+    if (g("O", "K") != "OK") return "Fail"
```
'Fail 2' and so on

sfs replied:
Done

max-kammerer commented on compiler/testData/codegen/bytecodeText/stringOperations/multipleNestedConcat.kt:
```diff
@@ -4,7 +4,7 @@
 fun test(s1: String, s2: String, s3: String): String {
     fun foo(s: String) = s
-    return "foo: " + foo(s1 + s2 + " ${foo("\${s3.length} = ${s3.length}")}")
+    return "foo: " + foo(s1 + s2 + " ${foo("\${s3.length} = ${s3.length} ")}") + " "
```
@sfs Why test update is required?

max-kammerer replied:
@sfs Oh..seems 'Intrinsics.stringPlus'. It's intresting that old backend uses 'Intrinsics.stringPlus' only if receiver is nullable, with receiver is non-nullable than StringBuilder is used directly. It's not clear which way is better. Is JVM differ these two cases on any string optimizations? //cc @udalov

dnpetrov replied:
HotSpot optimizes string concatenation (`"a" + "b" + "c"`, which is, effectively, `new StringBuilder().append("a").append("b").append("c")`).

max-kammerer replied:
@dnpetrov Yes pr does same, 
the question is 
```
new StringBuilder
dup
init
aload var1
append
aload var2 or LDC constant
append 
toString
```
would be same as efficient on Hotspot  and other JVMs as
```
aload var1
aload var2 or LDC constant
invoke Intrinsics.stringPlus

fun Intrinsics.stringPlus(var1, var2)
new StringBuilder
dup
init
aload var1
append
aload var2
append 
toString
```

sfs replied:
Having a single stringPlus call is slightly less code than creating a new StringBuilder.

If there are no plans to remove the stringPlus intrinsic then we might as well use it. I don't think there would be any performance difference in practice, since Intrinsic.stringPlus is a small leaf function which should be inlined (when hot) and uses StringBuilder.append internally.

max-kammerer replied:
@sfs Yes it should be inlined, but after inlining it could be a bit different bytecode, so if some pattern based string optimization is used further...

sfs replied:
@max-kammerer I can't give a definite answer to this, but as far as I can tell the StringBuilder related optimizations (StringOpts) run [as part of inlining](https://github.com/sourcemirror/jdk-8-hotspot/blob/af290d0fb302ee6a355fc55bc11602fddbac08bb/src/share/vm/opto/compile.cpp#L2013).

max-kammerer commented:
@sfs Thank you!

udalov commented on compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/FlattenStringConcatenationLowering.kt:
```diff
@@ -73,11 +75,13 @@ class FlattenStringConcatenationLowering(val context: CommonBackendContext) : Fi
+                    val receiver = expression.dispatchReceiver ?: expression.extensionReceiver
+                    receiver != null &&
+                            receiver.type.isStringClassType() &&
+                            function.getPackageFragment()?.fqName == KotlinBuiltIns.BUILT_INS_PACKAGE_FQ_NAME &&
```
@sfs Please use `parent` directly and check that it's a package fragment instead of using `getPackageFragment` which walks up until it finds a package fragment (which is not precise enough and _could_ result in incorrect behavior here for a member function, but it seems impossible due to other conditions).

sfs replied:
@udalov That was actually on purpose, since there are two functions we check for, `kotlin.String.plus` for non-nullable strings and `kotlin.plus` for nullable strings. If you prefer I could instead check for these two cases explicitly.

udalov replied:
Thanks for the clarification. Yes, I'd prefer that since it's more explicit and avoids confusion with other functions named `plus` in the standard library (e.g. if `receiver`'s initializer is refactored to `expression.extensionReceiver ?: expression.dispatchReceiver`, we could incorrectly identify some member extension on `String` named `plus` in an irrelevant class here)
